man/sawfish-client.1
.TH "sawfish-client" "1" 
.SH "NAME" 
sawfish-client \(em Evaluate arbitrary Lisp forms with sawfish. 
.SH "SYNOPSIS" 
.PP 
\fBsawfish-client\fR [\fB \-\-display \fIX \fR\fP]  [\fB \-q \fP]  [\fB \-f \fIFUNCTION \fR\fP]  [\fB \-c \fICOMMAND \fR\fP]  [\fB \-r \fIFEATURE \fR\fP]  [\fB \-e \fIFORM \fR\fP]  [\fB \- \fP]  [\fB \-\- \fP]  
.SH "DESCRIPTION" 
.PP 
This allows you to connect to a window manager process and 
evaluate arbitrary Lisp forms. 
.SH "OPTIONS" 
.IP "\fB\-\-display\fP" 10 
Connect to the window manager on display X 
.IP "\fB-q\fP" 10 
Be quiet (perform commands asynchronously) 
.IP "\fB-f\fP 	  \fBFUNCTION\fP 	" 10 
Call Lisp function \fIFUNCTION\fR on the server 
.IP "\fB-c\fP 	  \fBCOMMAND\fP 	" 10 
Call the interactive Lisp function \fICOMMAND\fR 
.IP "\fB-r\fP 	  \fBFEATURE\fP 	" 10 
Require the module called \fIFEATURE\fR 
.IP "\fB-e\fP 	  \fBFORM\fP 	" 10 
Evaluate Lisp form \fIFORM\fR on the server 
.IP "\fB-\fP" 10 
Read lines of input until \fIEOF\fR, evaluating each one as it is read 
.IP "\fB\-\-\fP" 10 
Read forms from standard input until \fIEOF\fR, evaluating the whole lot in one go (inside a progn) 
.IP "\fB\-\-help\fP" 10
Prints help.
.SH "SEE ALSO" 
.PP 
Sawfish are documented fully by \fIJohn Harper\fP       available via the \fBInfo\fP system . 
.SH "AUTHOR" 
.PP 
This manual page was written by Christian Marillat marillat@debian.org for 
the \fBDebian GNU/Linux\fP system (but may be used by others). 
.\" created by instant / docbook-to-man, Sat 02 Feb 2008, 23:15 


ChangeLog
2010-06-22  Christopher Bratusek <zanghar@freenet.de>
	* lisp/sawfish/wm/menus.jl: add a separator to the start of the root menu
	                            the menu might disappear directly due to the
				    dummy "Sawfish Rootmenu" headline, this border
				    prevents that.

2010-06-22  Teika kazura <teika@lavabit.com>
	* man/news.texi
	* src/functions.c: x-atom-name accepts full 32 unsigned
	int. Previously only 30 bits are treated, if the architecture is
	32-bit.

	* lisp/sawfish/wm/server.jl
	* man/news.texi: Prints backtrace for sawfish-client, too.
	This makes sense if you're using librep >= 0.90.7. Harmless anyway.

	* lisp/sawfish/cfg/main.jl.in
	* lisp/sawfish/cfg/shell.jl
	* man/news.texi: Made `sawfish-config' minimal, moving most the
	contents to sawfish.cfg.shell.  This facilitates hacks.

	* lisp/sawfish/wm/ext/error-handler.jl
	* man/news.texi: Better error message.
	Phrase "Sawfish error:" is prepended to error messages printed to
	standard error.

	* lisp/sawfish/wm/util/compat.jl
	* man/news.texi: Restored two options from "obsolete" declaration.
	They're workspace-boundary-mode and workspace-send-boundary-mode.

	* lisp/sawfish/wm/commands.jl
	* man/faq.texi
	* man/news.texi: Manuals (1). All options are now described.

	* CONTRIBUTING
	* lisp/sawfish/wm/commands.jl
	* lisp/sawfish/wm/ext/old-window-menu.jl
	* man/faq.texi
	* man/news.texi
	* man/sawfish-client.1
	* man/sawfish.texi
	* src/display.c
	* src/libclient.h: Doc.
	Improvement of news and man on the new "border" feature.
	Two new faq items.
	Added Daniel M. German to "thanks" list.

	* sawfish.ebuild.in: Declared that it is not an official one of
	any distribution.

	* sawfish-wm.desktop: Encoding fix in sawfish-wm.desktop.
	Greek was broken.

	* lisp/sawfish/wm/autoload.jl
	* lisp/sawfish/wm/ext/3d-hack.jl: Not autoloaded any more.

2010-05-28  Jeremy Hankins <nowan@nowan.org>
	* lisp/sawfish/wm/util/prompt.jl: Fix compilation error
	                 -- [Harald van Dijk]

2010-05-23  Jeremy Hankins <nowan@nowan.org>
	* lisp/sawfish/wm/util/prompt-extras.jl
	* lisp/sawfish/wm/util/prompt-wm.jl
	* lisp/sawfish/wm/util/prompt.jl: Changed the way prompt is called:
		helper functions are now passed directly instead of implicitly
		via let statements.  Also reverts the previous compilation error
		fix, since it's no longer necessary.

2010-05-23  Christopher Roy Bratusek <zanghar@freenet.de>
	* man/sawfish.texi
	* src/events.c
	* src/frames.c
	* src/sawfish.h
	* src/windows.c: added border_with and border_color frame-part attributes
			 -- [Alexey I. Froloff]

2010-05-16  Christopher Bratusek <zanghar@freenet.de>
	* src/display.c: allow compilation with -DDEBUG

	* man/news.texi: updated

2010-05-15  Teika Kazura <teika@lavabit.com>
	* man/news.texi: Added a link.

2010-05-11  Teika Kazura <teika@lavabit.com>
	* man/news.texi
	* man/sawfish.texi
	* src/display.c
	* src/events.c
	* src/functions.c
	* src/sawfish_subrs.h: Complets the last RandR edit. At least it's
	correct if you have only one screen, or if you use Xinerama.
	Most importantly, screen size information is updated after
	XRRScreenChangeNotifyEvent.
	Other changes:
	NULL display check of the RandR is changed to Sawfish conventional
	way. (Just to be in accordance.) Minor memeory leak fix.
	Doc.

	* ChangeLog
	* man/faq.texi
	* man/news.texi
	* man/sawfish.texi
	* KEYBINDINGS
	* src/keys.c: docs.

2010-04-27  Christopher Bratusek <zanghar@freenet.de>
	* configure.in: make xinerama an optional dependency
			-- [modified patch from Gentoo]

	* lisp/sawfish/wm/ext/apps-menu.jl: Allow whitespace and comments
	                                    at the beginning of .desktop files.
					    -- [Matthew Love, Teika Kazura]

	* src/Makefile.in
	* src/display.c
	* src/events.c: add a way to response to XRandR changed events
	                eg resolution changes via randr-change-notify-hook
			-- [Daniel M. German]

2010-04-06  Teika Kazura <teika@lavabit.com>
	* lisp/sawfish/wm/autoload.jl
	* lisp/sawfish/wm/commands.jl
	* lisp/sawfish/wm/util/gtkrc.jl
	* man/news.texi
	* man/sawfish.texi: Minor doc. News improvements. Minor typo fix.

	* lisp/sawfish/wm/ext/old-window-menu.jl: Utf8 support in
	old-window-menu.jl.

2010-03-20  Christopher Bratusek <zanghar@freenet.de>
	* lisp/sawfish/wm/tabs/tab.jl
	* lisp/sawfish/wm/tabs/tabgroup.jl: several additions, check man/news.texi [Fuchur]

	* man/news.texi: updated

2010-03-20  Teika Kazura <teika@lavabit.com>
	* lisp/sawfish/wm/windows.jl
	* man/news.texi
	* man/sawfish.texi: Implemented 'spec' option for window-class.

	* lisp/sawfish/wm/windows.jl
	* man/news.texi
	* man/sawfish.texi: Implemented 'window-pid'. Checked that even if
	pid is not available it doesn't signal an error (with xv).

	* README
	* man/news.texi
	* man/sawfish.texi
	* src/keys.c: Doc. Info "Events" chapter improvements. News revision.

2010-03-19  Christopher Bratusek <zanghar@freenet.de>
	* lisp/sawfish/wm/state/maximize.jl: make maximize-/-fullscreen/fullxinerama
	                                     honor maximize-raises [Nolan Leake]

	* lisp/sawfish/wm/ext/apps-menu.jl: Anti-brekage for "R".
	Removal of phony constants. Some code refactoring. [Teika Kazura]

2010-03-13  Christopher Bratusek <zanghar@freenet.de>
	* lisp/sawfish/wm/util/display-wininfo.jl: missing require

2010-03-13  Teika Kazura <teika@lavabit.com>
	* lisp/sawfish/wm/util/display-wininfo.jl
	* man/news.texi
	* man/sawfish.texi: 'display-window-position' command is implemented.

	* lisp/sawfish/wm/state/maximize.jl: Deleted bogus exports.

	* lisp/sawfish/wm/custom.jl
	* lisp/sawfish/wm/stacking.jl: Fix of a semi-bug. (defgroup stacking)
	is moved to custom.jl from stacking. Without this, adding 'tab' to
	open clause of wm.frames breaks the startup. Read git commit
	928355 log for the details.

	* README
	* lisp/sawfish/wm/ext/old-window-menu.jl: Minor doc change.

2010-03-06  Teika Kazura <teika@lavabit.com>
	* OPTIONS
	* lisp/sawfish/wm/util/display-window.jl

scripts/sawfish-client.jl
#!/bin/sh
exec rep "$0" "$@"
!#

;; sawfish-client.jl -- window manager remote client
;;
;; $Id$
;;
;; Copyright (C) 2000 John Harper <john@dcs.warwick.ac.uk>
;;
;; This file is part of sawmill.
;;
;; sawmill is free software; you can redistribute it and/or modify it
;; under the terms of the GNU General Public License as published by
;; the Free Software Foundation; either version 2, or (at your option)
;; any later version.
;;
;; sawmill is distributed in the hope that it will be useful, but
;; WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.
;;
;; You should have received a copy of the GNU General Public License
;; along with sawmill; see the file COPYING.  If not, write to
;; the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.

(require 'sawfish.client)

(define current-repl (make-fluid))

(define (usage)
  (write standard-output "\
usage: sawfish-client OPTIONS...

where OPTIONS are any of:

        --display X     Connect to the window manager on display X
        -q              Be quiet (perform commands asynchronously)
        -f FUNCTION     Call Lisp function FUNCTION on the server
        -c COMMAND      Call the interactive Lisp function COMMAND
        -r FEATURE      Require the module called FEATURE
        -e FORM         Evaluate Lisp form FORM on the server
        -               Read lines of input until EOF, evaluating each
                         one as it is read
        --              Read forms from standard input until EOF, evaluating
                         the whole lot in one go (inside a progn)\n"))

(define (exit n) (throw 'quit n))

(define (main)
  (let ((dpy (or (get-command-line-option "-display" t)
                 (get-command-line-option "--display" t))))
    (when dpy
      (sawfish-client-display dpy)))

  (when (get-command-line-option "--help")
    (usage)
    (exit 0))

  (let ((quiet nil))

    (define (do-eval form)
      (if quiet
          (sawfish-client-eval-async form)
        (write standard-output (sawfish-client-eval form t))
        (write standard-output #\newline)))

    (let loop ((args (or command-line-args '("-"))))
      (cond ((null args))

            ((string= (car args) "-q")
             (setq quiet t)
             (loop (cdr args)))

            ((string= (car args) "-d") (loop (cdr args)))

            ((and (string= (car args) "-f") (cdr args))
             (do-eval `(,(read-from-string (cadr args))))
             (loop (cddr args)))

	    ((and (string= (car args) "-c") (cdr args))
	     (do-eval `(call-command ',(read-from-string (cadr args))))
	     (loop (cddr args)))

            ((and (string= (car args) "-r") (cdr args))
	     (do-eval `(require ',(read-from-string (cadr args))))
	     (loop (cddr args)))

	    ((and (string= (car args) "-e") (cdr args))
	     (do-eval (read-from-string (cadr args)))
	     (loop (cddr args)))

	    ((string= (car args) "-")
	     (require 'rep.io.readline)
	     (require 'rep.util.repl)
	     (format standard-output "\
sawfish %s, Copyright (C) 1999-2000 John Harper
sawfish comes with ABSOLUTELY NO WARRANTY; for details see the file COPYING\n"
		     (sawfish-client-eval 'sawfish-version))
	     (let ((r (sawfish-client-eval
		       `(progn
			  (require 'rep.util.repl)
                          (define-repl-command 'quit
                            (lambda () (throw 'sawfish-client-exit)))
			  (make-repl 'user)))))
	       (let-fluids ((current-repl r))
		 (write standard-output "\nEnter `,help' to list commands.\n")
		 (let loop ()
		   (let ((input (readline
				 (format nil (if (repl-pending r) "" "%s> ")
					 (repl-struct r)))))
		     (when input
		       (let ((out (sawfish-client-eval
				   `(progn
				      (require 'rep.util.repl)
				      (let* ((standard-output
					      (make-string-output-stream))
					     (standard-error standard-output)
					     (r ',r))
					(cons (catch 'sawfish-client-exit
                                                (and (repl-iterate r ',input)
                                                     r))
					      (get-output-stream-string
					       standard-output)))))))
			 (setq r (car out))
			 (fluid-set current-repl r)
			 (write standard-output (cdr out))
			 (when r (loop))))))))
	     (loop (cdr args)))

	    ((string= (car args) "--")
	     (let ((text (make-string-output-stream)))
	       (copy-stream standard-input text)
	       (do-eval (read-from-string
			 (concat "\(progn"
				 (get-output-stream-string text)
				 "\)")))
	       (loop (cdr args))))

	    (t (format standard-error
		       "sawfish-client: unknown option `%s'\n" (car args))
	       (format standard-error
		       "Try `sawfish-client --help' for more information.\n")
	       (exit 1))))
    (exit 0)))

(define (rl-completion-generator w)
  (sawfish-client-eval
   `(repl-completions ',(fluid current-repl) ',w)))

(main)

;; Local variables:
;; major-mode: lisp-mode
;; End:
